tests/unit/test_config_provider.py

```diff
@@ -942,3 +943,174 @@ def test_resolve_auto_mode_imds_region_provider_connect_timeout(self):
+            },
+            full_config_map={},
+            scoped_config_map={},
+            expected_value='global-endpoint-url',
+        )
+
+    def test_can_provide_global_config_var(self):
+        full_config_map = {
+            'profiles': {'default': {'endpoint_url': 'global-from-config'}}
+        }
+
+        scoped_config_map = full_config_map['profiles']['default']
+
+        self.assert_does_provide(
+            service="ec2",
+            environ_map={},
+            full_config_map=full_config_map,
+            scoped_config_map=scoped_config_map,
+            expected_value='global-from-config',
+        )
+
+    def test_can_provide_service_config_var(self):
+        full_config_map = {
+            'profiles': {'default': {'services': 'my-services'}},
+            'services': {
+                'my-services': {
+                    'batch': {'endpoint_url': 'global-from-config'}
+                }
+            },
+        }
+
+        scoped_config_map = full_config_map['profiles']['default']
+
+        self.assert_does_provide(
+            service="batch",
+            environ_map={},
+            full_config_map=full_config_map,
+            scoped_config_map=scoped_config_map,
+            expected_value='global-from-config',
+        )
+
+    def test_can_provide_service_config_var_over_global(self):
+        full_config_map = {
+            'profiles': {
+                'default': {
+                    'services': 'my-services',
+                    'endpoint_url': 'global-from-config',
+                }
+            },
+            'services': {
+                'my-services': {'batch': {'endpoint_url': "batch-from-config"}}
+            },
+        }
+
+        scoped_config_map = full_config_map['profiles']['default']
+        self.assert_does_provide(
+            service="batch",
+            environ_map={},
+            full_config_map=full_config_map,
+            scoped_config_map=scoped_config_map,
+            expected_value='batch-from-config',
+        )
+
+    def test_can_provide_service_config_var_over_global_diff_service(self):
+        full_config_map = {
+            'profiles': {
+                'default': {
+                    'services': 'my-services',
+                    'endpoint_url': 'global-from-config',
+                }
+            },
+            'services': {
+                'my-services': {'batch': {'endpoint_url': "batch-from-config"}}
+            },
+        }
+
+        scoped_config_map = full_config_map['profiles']['default']
+
+        self.assert_does_provide(
+            service="s3",
+            environ_map={},
+            full_config_map=full_config_map,
+            scoped_config_map=scoped_config_map,
+            expected_value='global-from-config',
+        )
+
+
+@pytest.mark.parametrize(
+    "service_name,service_id", _known_service_names_and_ids()
+)
+def test_service_env_var_name_is_correct(service_name, service_id):
```

Comment:
Move this to a functional test and use a session, etc. to test it end-to-end with all services.